source/priority_queue.c

```diff
@@ -0,0 +1,135 @@
+    int left = (index << 1) + 1;
+    int right = left + 1;
+    int swap_candidate = index;
+    int len = aws_array_list_length(&queue->container);
+
+
+    void *item, *swap_item;
+    aws_array_list_get_at_ptr(&queue->container, &item, left);
+    aws_array_list_get_at_ptr(&queue->container, &swap_item, swap_candidate);
+
+    if (left < len && queue->compare(item, swap_item)) {
+        swap_candidate = left;
+        swap_item = item;
+    }
+
+    aws_array_list_get_at_ptr(&queue->container, &item, right);
+    if (right < len && queue->compare(item, swap_item)) {
+        swap_candidate = right;
+    }
+
+    if (swap_candidate != index) {
+        aws_array_list_swap(&queue->container, index, swap_candidate);
+        make_heap_from_top(queue, swap_candidate);
+    }
+}
+
+static void make_heap_from_bottom(struct aws_priority_queue *queue, int index) {
+    int parent = 0;
+
+    if (index & 0x01) {
+        parent = (index - 1) >> 1;
+    }
+    else {
+        parent = (index - 2) >> 1;
+    }
+
+    int swap_candidate = index;
+
+    void *parent_item, *swap_item;
+    aws_array_list_get_at_ptr(&queue->container, &parent_item, parent);
+    aws_array_list_get_at_ptr(&queue->container, &swap_item, swap_candidate);
+
+    if (parent >= 0 && queue->compare(swap_item, parent_item)) {
+        swap_candidate = parent;
+    }
+
+    if (swap_candidate != index) {
+        aws_array_list_swap(&queue->container, index, swap_candidate);
+        make_heap_from_bottom(queue, swap_candidate);
+    }
+}
+
+int aws_priority_queue_dynamic_init(struct aws_priority_queue *queue, struct aws_allocator *alloc, size_t default_size,
+                                    size_t item_size, int(*compare)(void *a, void *b)) {
+
+    queue->compare = compare;
+    return aws_array_list_init_dynamic(&queue->container, alloc, default_size, item_size);
+}
+
+void aws_priority_queue_static_init(struct aws_priority_queue *queue,
+    void *heap, size_t item_count, size_t item_size, int(*compare)(void *a, void *b)) {
+
+    queue->compare = compare;
+    aws_array_list_init_static(&queue->container, heap, item_count, item_size);
+}
+
+void aws_priority_queue_clean_up(struct aws_priority_queue *queue) {
+    aws_array_list_clean_up(&queue->container);
+}
+
+int aws_priority_queue_push(struct aws_priority_queue *queue, void* item) {
+    int err = aws_array_list_push_back(&queue->container, item);
+    if(err) {
+        return err;
+    }
+
+    make_heap_from_bottom(queue, aws_array_list_length(&queue->container) - 1);
+
+    return AWS_OP_SUCCESS;
+}
+
+int aws_priority_queue_pop(struct aws_priority_queue *queue, void* item) {
+
+    if(0 == aws_array_list_length(&queue->container)) {
+        return AWS_ERROR_PRIORITY_QUEUE_EMPTY;
+    }
+
+    int err = aws_priority_queue_top(queue, item);
+
+    if (err) {
+        return err;
+    }
+
+    err = aws_array_list_pop_front(&queue->container);
```

Comment:
This will shift all elements down one, which is not what you want when doing a heap as it'll effectively destroy the entire heap tree structure. You need to swap in the element currently at the end, then fix the heap with a down-heap operation.

Reply:
doh! good catch.